### PR TITLE
0115 ct mappability

### PR DESCRIPTION
**Pull request by ramiromagno**
Thank you for your Pull Request! We have developed this task checklist from the
[Development Process
Guide](https://pharmaverse.github.io/sdtm.oak/CONTRIBUTING.html)
to help with the final steps of the process. Completing the below tasks helps to
ensure our reviewers can maximize their time on your code as well as making sure
the oak codebase remains robust and consistent.

Please check off each taskbox as an acknowledgment that you completed the task
or check off that it is not relevant to your Pull Request. This checklist is
part of the Github Action workflows and the Pull Request will not be merged into
the `devel` branch until you have checked off each task.

- [x] Place Closes #<insert_issue_number> into the beginning of your Pull z
Request Title (Use Edit button in top-right if you need to update)
- [x] Code is formatted according to the 
[tidyverse style guide](https://style.tidyverse.org/). Run 
`styler::style_file()` to style R and Rmd files
- [x] Updated relevant unit tests or have written new unit tests, which should
consider realistic data scenarios and edge cases, e.g. empty datasets, errors,
boundary cases etc. - See 
[Unit Test Guide](https://pharmaverse.github.io/admiraldev/articles/unit_test_guidance.html#tests-should-be-robust-to-cover-realistic-data-scenarios)
- [x] If you removed/replaced any function and/or function parameters, did you 
fully follow the 
[deprecation guidance](https://pharmaverse.github.io/admiraldev/articles/programming_strategy.html#deprecation)?
- [x] Update to all relevant roxygen headers and examples, including keywords
and families. Refer to the 
[categorization of functions](https://pharmaverse.github.io/admiraldev/articles/programming_strategy.html#categorization-of-functions) to tag appropriate keyword/family.
- [x] Run `devtools::document()` so all `.Rd` files in the `man` folder and the
`NAMESPACE` file in the project root are updated appropriately
- [x] Address any updates needed for vignettes and/or templates
- [x] Update `NEWS.md` if the changes pertain to a user-facing function (i.e. it
has an `@export` tag) or documentation aimed at users (rather than developers)
- [x] Build oak site `pkgdown::build_site()` and check that all affected
examples are displayed correctly and that all new functions occur on the "[Reference](https://pharmaverse.github.io/admiral/reference/index.html)" page. 
- [x] Address or fix all lintr warnings and errors - `lintr::lint_package()`
- [x] Run `R CMD check` locally and address all errors and warnings - `devtools::check()`
- [x] Link the issue in the Development Section on the right hand side.
- [x] Address all merge conflicts and resolve appropriately
- [x] Pat yourself on the back for a job well done! Much love to your accomplishment!

### CODE REVIEW

**Comment by ShiyuC** [comment]
@ramiromagno thanks for adding the new function. I tested on my end and the message printed out nicely. Could you please add a testcase in the package for this change?

**Comment by ramiromagno** [comment]
@ShiyuC : Nice! Could you provide a nice, simple and meaningful example? I'm happy to add it then as a unit test.

**Comment by ShiyuC** [comment]
@ramiromagno I used the VS domain template we have in the package, and just updated the first function by changing the `tgt_val` to "abc", and the message is popped up.
image

**Comment by ramiromagno** [comment]
okay, thanks, but in that case I think I will go with an example we have in docs.

**Review by ShiyuC — APPROVED** [review]
Looks good to me 💯

File: R/ct.R [hunk 1]
```diff
@@ -184,6 +184,56 @@ ct_mappings <- function(ct_spec, from = ct_spec_vars("from"), to = ct_spec_vars(
...
+#' @keywords internal
+inform_on_ct_mappability <- function(x, from) {
+  is_mappable <- is_ct_mappable(x, from)
+  if (all(is_mappable)) {
+    return(invisible(TRUE))
+  }
+
+  unmappable <- unique(x[!is_mappable])
+
+  cli::cli_alert_info("These values could not be mapped to controlled terminology: {.val {unmappable}}.",
```
Comment (rammprasad) [inline R/ct.R:230]: Can we change the message as `These values could not be mapped as per the controlled terminology: {.val {unmappable}}. Update the controlled terminology file to include this value.`

**Review by rammprasad — CHANGES_REQUESTED** [review]
Added my comments.

File: R/ct.R [hunk 2]
```diff
@@ -184,6 +184,56 @@ ct_mappings <- function(ct_spec, from = ct_spec_vars("from"), to = ct_spec_vars(
...
+#' @param x A character vector of terms to be evaluated for recoding.
+#' @param from A character vector of controlled terminology terms that `x` will
+#'   be compared against.
+#'
+#' @return A logical vector of the same length as `x`, where `TRUE` indicates the
+#'   corresponding term in `x` is found in `from`, and `FALSE` otherwise.
+#'
+#' @keywords internal
+is_ct_mappable <- function(x, from) {
+  !is.na(index_for_recode(x = x, from = from))
```
[the remark this answers is not in the document]
Comment (ramiromagno) [reply R/ct.R:206]: I'm sorry but that feels like an anti-pattern. I think we should alert the user on trying to hardcode an empty string with controlled terminology. If setting an empty string is the user's real intent, then, perhaps they should be redirected to using `hardcode_no_ct()` instead.

File: R/ct.R (hunk 1 above)
Comment (ramiromagno) [reply R/ct.R:230]: I would prefer a terser message, as this can happen quite a few time throughout a domain derivation, leading to spamming the R console with lots of textual repetition. And I think that the course-of-action recommendation should be more neutral with regards to what went wrong. It could be that there was a typo in the hardcoded value, not an incomplete CT per se. Also, we should refer to the controlled terminology in a more abstract way, as it does not need necessarily to come from a file upstream (e.g. it might come from an external package).

File: R/ct.R (hunk 2 above)
[the remark this answers is not in the document]
Comment (ramiromagno) [reply R/ct.R]: I believe this is solved now. Let me know otherwise.

File: R/ct.R (hunk 1 above)
Comment (ramiromagno) [reply R/ct.R]: Updated the message.

**Review by rammprasad — APPROVED** [review]
LGTM
